File: models/nerModel.js
const nlp = require('compromise');

exports.identifyEntities = async (sbom) => {
    const sbomText = JSON.stringify(sbom);
    const doc = nlp(sbomText);
    const entities = [];



    doc.match('#Software').json().forEach(software => {
        entities.push({ entity: 'Software', value: software.text });
    });

    doc.match('#Dependency').json().forEach(dependency => {
        entities.push({ entity: 'Dependency', value: dependency.text });
    });

    doc.match('#Vulnerability').json().forEach(vulnerability => {
        entities.push({ entity: 'Vulnerability', value: vulnerability.text });
    });

    doc.match('#Metadata').json().forEach(metadata => {
        entities.push({ entity: 'Metadata', value: metadata.text });
    });

    return entities;
};

File: backend/models/ner_model.py
import spacy

nlp = None

def load_model():
    global nlp
    if nlp is None:
        nlp = spacy.load("en_core_web_sm")

def identify_entities(sbom):
    load_model()
    sbom_text = str(sbom)
    doc = nlp(sbom_text)
    entities = []
    for ent in doc.ents:
        if ent.label_ == "ORG":
            entities.append({"entity": "Software", "value": ent.text})
        elif ent.label_ == "PRODUCT":
            entities.append({"entity": "Dependency", "value": ent.text})
        elif ent.label_ == "NORP":
            entities.append({"entity": "Vulnerability", "value": ent.text})
        elif ent.label_ == "MISC":
            entities.append({"entity": "Metadata", "value": ent.text})
    return entities
